src/sentry/search/django/backend.py

```diff
@@ -250,6 +257,80 @@ def query(self, project, tags=None, environment=None, sort_by='date', limit=100,
         else:
             retention_window_start = None
 
+        # TODO: This could also do something janky like switching based on the current
+        #       tagstore backend, or a global option, or a number of other things. Going
+        #       by the project seemed the best for now.
+        use_snuba = bool(project.get_option('sentry:snuba', False))
+        if use_snuba and (tags or
+                          environment is not None or
+                          any(key in parameters for key in ('date_from', 'date_to'))):
+
+            # if we're in this branch then Snuba can help us filter down the eligible group ids
+            filters = {
+                'project_id': [project.id],
+            }
+
+            if environment is not None:
+                filters['environment'] = [environment.id]
+
+            conditions = []
+            for tag, val in six.iteritems(tags):
+                col = 'tags[{}]'.format(tag)
+                if val == ANY:
+                    conditions.append((col, 'IS NOT NULL', None))
+                else:
+                    conditions.append((col, '=', val))
+
+            now = timezone.now()
+            end = parameters.get('date_to') or now
+            # TODO: Presumably we want to search back to the project's full retention,
+            #       which may be higher than 90 days in the future, but apparently
+            #       `retention_window_start` can be None?
+            start = (parameters.get('date_from')
+                     or retention_window_start
+                     or (now - timedelta(days=90)))
+
+            hashes = snuba.query(start, end, ['primary_hash'], conditions, filters)
+
+            group_ids = GroupHash.objects.filter(
+                project=project, hash__in=hashes.keys()
+            ).values_list('group_id', flat=True).distinct()
+            group_queryset = group_queryset.filter(id__in=group_ids)
+
+            # The following isn't different from the original Django-only else branch
+            # below, except tags are handled above in Snuba code. From this point on
+            # we're just working with the `group_queryset` that has been filtered down
+            # to a set of `Group.id`s by Snuba.
```

Comment:
It'd be nice to avoid duplicating this if possible, but I haven't thought through it enough to make a concrete suggestion on how to do that.

Reply:
I feel the same but I think the idea is that this is short (single digit months...?) lived?

Reply:
Cleaning this up as part of making `SnubaSearch` backend.